### PR TITLE
Adds new metrics for some apollo specific env variables

### LINKED ISSUE #4290: Offline license - Operation metrics
Add the following env metrics:

* apollo.router.env - gauge for env variables
    * env.apollo.key=true
    * env.apollo.graph_ref=true
    * env.apollo.license=true

* apollo.router.license - gauge for license state
    * license.warn = true
    * license.error = true

### PR DESCRIPTION
**Pull request by BrynCooke**
Adds some metrics for some specific apollo env variables to show if they are set or not.
Uses the new metrics macros rather than the hash map to exercise more of the stack.
As a bonus, the thread that was used to continuously update the gauges is gone. 

I had to make some changes to the asserting to allow asserting of all metrics rather than just specific ones hence all the snapshot changes. On the plus side they look better now.

Fixes #4290

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-router/src/configuration/metrics.rs [hunk 1]
```diff
@@ -336,20 +321,74 @@ impl Metrics {
...
+impl From<InstrumentData> for Metrics {
+    fn from(data: InstrumentData) -> Self {
+        Metrics {
+            _instruments: data
+                .data
+                .into_iter()
+                .map(|(metric_name, (value, attributes))| {
+                    let attributes: Vec<_> = attributes
+                        .into_iter()
+                        .map(|(k, v)| KeyValue::new(k.replace("__", "."), v.clone()))
```
Comment (o0Ignition0o) [inline apollo-router/src/configuration/metrics.rs:373]:
```suggestion
                        .map(|(k, v)| KeyValue::new(k.replace("__", "."), v))
```

is the clone necessary here?

File: apollo-router/src/configuration/metrics.rs [hunk 2]
```diff
@@ -336,20 +321,74 @@ impl Metrics {
...
+                        .collect();
+                    data.meter
+                        .u64_observable_gauge(metric_name)
+                        .with_callback(move |observer| {
+                            observer.observe(value, &attributes);
```
Comment (o0Ignition0o) [inline apollo-router/src/configuration/metrics.rs:378]: it s a pity oserve doesn't take an iterator :(

File: apollo-router/src/configuration/metrics.rs [hunk 3]
```diff
@@ -1,150 +1,139 @@
...
-                // Now log the metric
-                tracing::info!($($metric).+ = metric.0);
-
             };
             ($($metric:ident).+, $path:literal, $($($attr:ident).+, $attr_path:literal),+) => {
-                let metric_name = stringify!($($metric).+).to_string();
-                let metric = self.metrics.entry(metric_name.clone()).or_insert_with(|| {
-                    if let Some(value) = JsonPathInst::from_str($path).expect("json path must be valid").find_slice(&self.yaml).first() {
+                let instrument_name = stringify!($($metric).+).to_string();
+                self.data.entry(instrument_name.clone()).or_insert_with(|| {
```
Comment (o0Ignition0o) [inline apollo-router/src/configuration/metrics.rs:130]: is this clone necessary ?

**Review by o0Ignition0o — APPROVED** [review]
Not sure I've understood everything, but lgtm overall.

File: apollo-router/src/configuration/metrics.rs (hunk 1 above)
Comment (BrynCooke) [reply apollo-router/src/configuration/metrics.rs]: [a4e3605](https://github.com/apollographql/router/pull/4364/commits/a4e36056787230e0182223ffb78d9126e362dca1)

File: apollo-router/src/configuration/metrics.rs (hunk 3 above)
Comment (BrynCooke) [reply apollo-router/src/configuration/metrics.rs]: [a4e3605](https://github.com/apollographql/router/pull/4364/commits/a4e36056787230e0182223ffb78d9126e362dca1)

**Pull request merged**